### PR TITLE
feat/add-folder

### PR DESCRIPTION
**Pull request by kaitoz11**
## help my life with:
- tests
- feature

### CODE REVIEW

File: src/Api/Controllers/FoldersController.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+using Application.Common.Models.Dtos.Physical;
+using Application.Folders.Commands.AddFolder;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Api.Controllers;
+
+public class FoldersController : ApiControllerBase
+{
+    [HttpPost]
+    public async Task<ActionResult<Result<FolderDto>>> AddFolders(AddFolderCommand command)
```
Comment (StarryFolf) [inline src/Api/Controllers/FoldersController.cs:11]: should have [FromBody]

Comment (kaitoz11) [reply src/Api/Controllers/FoldersController.cs:11]: oops....

File: src/Api/Controllers/FoldersController.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+public class FoldersController : ApiControllerBase
+{
+    [HttpPost]
+    [ProducesResponseType(StatusCodes.Status200OK)]    
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(StatusCodes.Status403Forbidden)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    [ProducesResponseType(StatusCodes.Status409Conflict)]
+    public async Task<ActionResult<Result<FolderDto>>> AddFolders([FromBody] AddFolderCommand command)
```
Comment (ChienNQuang) [inline src/Api/Controllers/FoldersController.cs:16]: Plural form?

File: src/Application/Folders/Commands/AddFolder/AddFolderCommand.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    public AddFolderCommandHandler(IApplicationDbContext context, IMapper mapper)
+    {
+        _context = context;
+        _mapper = mapper;
+    }
+
+    public async Task<FolderDto> Handle(AddFolderCommand request, CancellationToken cancellationToken)
+    {
+        var locker = await _context.Lockers
+            .FirstOrDefaultAsync(l =>
```
Comment (ChienNQuang) [inline src/Application/Folders/Commands/AddFolder/AddFolderCommand.cs:34]: spacing?

File: src/Application/Folders/Commands/AddFolder/AddFolderCommandValidator.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+using FluentValidation;
+
+namespace Application.Folders.Commands.AddFolder;
+
+public class AddFolderCommandValidator : AbstractValidator<AddFolderCommand>
+{
+    public AddFolderCommandValidator()
+    {
+
```
Comment (ChienNQuang) [inline src/Application/Folders/Commands/AddFolder/AddFolderCommandValidator.cs:9]: spacing?

File: src/Application/Folders/Commands/AddFolder/AddFolderCommandValidator.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        RuleLevelCascadeMode = CascadeMode.Stop;
+
+        RuleFor(f => f.Name)
+            .NotEmpty().WithMessage("Name is required.")
+            .MaximumLength(64).WithMessage("Name cannot exceed 64 characters.");
+
+        RuleFor(f => f.Description)
+            .MaximumLength(256).WithMessage("Description cannot exceed 256 characters.");
+
+        RuleFor(f => f.Capacity)
```
Comment (ChienNQuang) [inline src/Application/Folders/Commands/AddFolder/AddFolderCommandValidator.cs:19]: This is required

File: tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        var folderEntity = await FindAsync<Folder>(folder.Id);
+        var lockerEntity = await FindAsync<Locker>(locker.Id);
+        var roomEntity = await FindAsync<Room>(room.Id);
+        Remove(folderEntity);
+        Remove(lockerEntity);
+        Remove(roomEntity);
+    }
+
+    [Fact]
+    public async Task ShouldAddFolder_WhenFoldersHasSameNameButInDifferentLockers()
```
Comment (ChienNQuang) [inline tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs:77]: This is not needed

File: tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+            Name = sameFolderName,
+            LockerId = lockerA.Id
+        };
+        var addFolderCommandForLockerB =  addFolderCommand with 
+        {
+            Name = sameFolderName,
+            LockerId = lockerB.Id
+        };
+        // Act
+        var folderA = await SendAsync(addFolderCommandForLockerA);
```
Comment (ChienNQuang) [inline tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs:118]: This should not be in Act

File: tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+        var folderB = await SendAsync(addFolderCommandForLockerB);
+        
+        // Assert
+        folderA.Locker.Id.Should().NotBe(folderB.Locker.Id);
+        folderA.Name.Should().Be(folderB.Name);
+        
+        folderB.Name.Should().Be(addFolderCommandForLockerB.Name);
```
Comment (ChienNQuang) [inline tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs:125]: These assertions are not needed

File: tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            Name = sameFolderName,
+            LockerId = locker.Id
+        };
+        var addFolderCommandForLockerB =  addFolderCommand with 
+        {
+            Name = sameFolderName,
+            LockerId = locker.Id
+        };
+        // Act
+        var folder = await SendAsync(addFolderCommandForLockerA);
```
Comment (ChienNQuang) [inline tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs:173]: This too should not be in Act

File: src/Api/Controllers/FoldersController.cs (hunk 2 above)
Comment (kaitoz11) [reply src/Api/Controllers/FoldersController.cs]: fixed at https://github.com/ProFile-org/ProFile-BE/pull/80/commits/6f136a5a8d36c9c5f790d5718e77bde1acd59818

File: src/Application/Folders/Commands/AddFolder/AddFolderCommand.cs (hunk 3 above)
Comment (kaitoz11) [reply src/Application/Folders/Commands/AddFolder/AddFolderCommand.cs]: fixed at https://github.com/ProFile-org/ProFile-BE/pull/80/commits/d87261e746bf0889e0651581bc05148e31e0eef9

File: src/Application/Folders/Commands/AddFolder/AddFolderCommandValidator.cs (hunk 4 above)
Comment (kaitoz11) [reply src/Application/Folders/Commands/AddFolder/AddFolderCommandValidator.cs:9]: fixed at https://github.com/ProFile-org/ProFile-BE/pull/80/commits/d87261e746bf0889e0651581bc05148e31e0eef9

File: tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs (hunk 6 above)
Comment (kaitoz11) [reply tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs:77]: yes it does

File: tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs (hunk 7 above)
Comment (kaitoz11) [reply tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs:117]: fixed at https://github.com/ProFile-org/ProFile-BE/pull/80/commits/001ea4b4f149a2b55bb60565c921ca2c54419550

File: tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs (hunk 8 above)
Comment (kaitoz11) [reply tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs]: fixed at https://github.com/ProFile-org/ProFile-BE/pull/80/commits/8aa67a16df679c621c6b8aaa0bf08384a795bbe1

File: tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs (hunk 9 above)
Comment (kaitoz11) [reply tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs:166]: fixed at https://github.com/ProFile-org/ProFile-BE/pull/80/commits/5bf85312dc8fdad26790b428b35e1a66ccad0130

File: tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+    {
+        // Arrange
+        var addFolderCommand = _folderGenerator.Generate();
+        addFolderCommand =  addFolderCommand with 
+        {
+            LockerId = Guid.NewGuid()
+        };
+        
+        // Act
+        var folder = async () => await SendAsync(addFolderCommand);
```
Comment (ChienNQuang) [inline tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs:241]: Spacing?

File: tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+            {
+                var addFolderCommand = _folderGenerator.Generate();
+                addFolderCommand = addFolderCommand with
+                {
+                    LockerId = locker.Id
+                };
+                list.Add(await SendAsync(addFolderCommand));    
+            }
+        };
+        // Assert
```
Comment (ChienNQuang) [inline tests/Application.Tests.Integration/Folders/Commands/AddFolderTests.cs:214]: Spacing?

**Review by ChienNQuang — APPROVED** [review]
LGTM

**Pull request merged**